src/Common/src/TypeSystem/Common/TypeSystemException.cs

```diff
@@ -0,0 +1,64 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+namespace Internal.TypeSystem
+{
+    /// <summary>
+    /// Base type for all type system exceptions.
+    /// </summary>
+    public abstract class TypeSystemException : Exception
+    {
+        public TypeSystemException(string message)
+            : base(message)
+        {
+        }
+
+        /// <summary>
+        /// The exception that is thrown when type-loading failures occur.
+        /// </summary>
+        public class TypeLoadException : TypeSystemException
+        {
+            internal TypeLoadException(string message)
+                : base(message)
+            {
+            }
+        }
+
+        /// <summary>
+        /// The exception that is thrown when there is an attempt to access a class member that does not exist
+        /// or that is not declared as public.
+        /// </summary>
+        public abstract class MissingMemberException : TypeSystemException
+        {
+            protected internal MissingMemberException(string message)
+                : base(message)
```

Comment:
I know that this is a problem, but I have trouble finding an easy way that wouldn't make this too annyoning. The problem is that we need the message both at runtime (from the generated throw helper) and at compile time (if we want to generate a warning or an error).

Ideal solution is to have some kind of scheme where we can reuse the helper that uses the message ID+parameters to get a formatted string that can be used from both places and reuse the message. This sounds like days of work fighting buildtools and it will likely become a nightmare for loc tooling too.

The other option (the annoying one) is to duplicate the messages and duplicate the formatter, but that means every time someone wants to throw a new type system exception, they need to add two messages, add the ID, and update two formatters :(.